tools/update.php

```diff
@@ -0,0 +1,890 @@
+{
+    // Backup source code to e.g. /tmp/bkp-LORIS_v19.x-dev_16-May-2018
+    $backup_dir      .= 'bkp-LORIS';
+    $version_filepath = $loris_root_dir . 'VERSION';
+    $backup_dir      .= '_v' . getVersionFromLORISRoot($loris_root_dir);
+    $backup_dir      .= '_' . date("j-M-Y") . '/'; // e.g. 10-May-2018
+    echo "[*] Backing up $loris_root_dir to $backup_dir" . PHP_EOL;
+    recurseCopy($loris_root_dir, $backup_dir);
+
+    // Get the release code from Github
+    $tarball_path = downloadLatestRelease();
+    if (empty($tarball_path)) {
+        die('[-] ERROR: Could not download the latest LORIS release.');
+    }
+    $dst_dir = '/tmp/'; // Parent directory for source code download
+    echo '[*] Extracting release files...' . PHP_EOL;
+    $cmd = "unzip -o " . escapeshellarg($tarball_path) . ' -d '
+        . escapeshellarg($dst_dir);
+    doExec($cmd);
+
+    // Find the file name for the release just downloaded
+    $release_dir = '';
+    foreach (glob("$dst_dir*") as $filename) {
+        if (strpos($filename, '/aces-Loris-') !== false) {
+            $release_dir = $filename;
+        }
+    }
+    if (empty($release_dir)) {
+        die("[-] ERROR: Could not find downloaded files in $dst_dir" . PHP_EOL);
+    }
+    // Use rsync to overwrite files in $loris_root
+    echo '[*] Overwriting old source code files.'  . PHP_EOL;
+    $cmd = 'rsync -r ' . escapeshellarg($release_dir) . ' ' .
+        escapeshellarg($loris_root_dir);
+    return doExec($cmd);
+}
+
+/**
+ * Ensures required packages and their repositories are known to the server
+ * and are up-to-date.
+ *
+ * @param array $requirements Apt packages required by LORIS.
+ *
+ * @return bool True if all packages installed and up-to-date. False otherwise.
+ */
+function updateRequiredPackages($requirements) : bool
+{
+    // we need 3rd party PPA to get the latest PHP on Ubuntu
+    echo '[*] Adding external PPAs...' . PHP_EOL;
+    // -y flag required to suppress a message from the author
+    // TODO: check if these already exist before adding
+    exec('sudo apt-add-repository ppa:ondrej/php -y');
+    exec('sudo apt-add-repository ppa:ondrej/apache2 -y');
+
+    echo '[*] Updating apt package list...' . PHP_EOL;
+    exec('sudo apt-get update');
+    // die unless all required packages are installed and up-to-date
+    if (!(installMissingRequirements($requirements))
+        && (installAptPackages($requirements, $upgrade_mode = true))
+    ) {
+        die(
+            'Could not upgrade all required packages. Exiting.'
+            . PHP_EOL
+        );
+    }
+    return true;
+}
+
+/**
+ * Returns a list of SQL patch files to apply based on LORIS version
+ * differences.
+ *
+ * @param string $loris_root   The directory where LORIS is installed.
+ * @param string $version_from Pre-update LORIS version.
+ * @param string $version_to   Post-update LORIS version.
+ *
+ * @return array Patch files to be applied. Empty if up-to-date.
```

Comment:
at this point I think this is just a single file

Reply:
Nope! For example if you're a developer and two patches have been released on the minor version but a new minor release patch hasn't been issued yet.